jvm-libs/linea/web3j-extensions/src/test/kotlin/linea/web3j/ethapi/Web3jEthApiClientWithRetriesTest.kt

```diff
@@ -0,0 +1,85 @@
+import kotlin.time.Duration.Companion.milliseconds
+
+@ExtendWith(VertxExtension::class)
+class Web3jEthApiClientWithRetriesTest {
+  private lateinit var jsonRpcServer: TestingJsonRpcServer
+  private lateinit var client: EthApiClient
+
+  @BeforeEach
+  fun setup(vertx: Vertx) {
+    jsonRpcServer = TestingJsonRpcServer(
+      vertx = vertx,
+      recordRequestsResponses = true,
+    )
+    client = createEthApiClient(
+      rpcUrl = jsonRpcServer.httpEndpoint.toString(),
+      requestRetryConfig = RetryConfig(maxRetries = 2u, backoffDelay = 10.milliseconds),
+      vertx = vertx,
+    )
+  }
+
+  @Test
+  fun `should retry until retries elapse`() {
+    jsonRpcServer.handle("eth_getBlockByNumber") { request ->
+      throw JsonRpcErrorResponseException(rpcErrorCode = -123, rpcErrorMessage = "Internal error")
+    }
+    runCatching { client.getBlockByNumberWithoutTransactionsData(BlockParameter.Tag.FINALIZED).get() }
+    assertThat(jsonRpcServer.recordedRequests()).hasSize(3)
+  }
+
+  @Test
+  fun `should stop retry when predicate is true`(vertx: Vertx) {
+    jsonRpcServer.handle("eth_getBlockByNumber") { request ->
+      throw JsonRpcErrorResponseException(rpcErrorCode = -39001, rpcErrorMessage = "Unknown block")
+    }
+    val client = createEthApiClient(
+      rpcUrl = jsonRpcServer.httpEndpoint.toString(),
+      requestRetryConfig = RetryConfig(maxRetries = 2u, backoffDelay = 10.milliseconds),
+      vertx = vertx,
+      stopRetriesOnErrorPredicate = { th ->
+        if (th.cause is linea.error.JsonRpcErrorResponseException) {
+          val rpcError = th.cause as linea.error.JsonRpcErrorResponseException
+          rpcError.method == "eth_getBlockByNumber" && rpcError.rpcErrorCode == -39001
+        } else {
+          false
+        }
+      },
+    )
+    runCatching { client.getBlockByNumberWithoutTransactionsData(BlockParameter.Tag.FINALIZED).get() }
+    assertThat(jsonRpcServer.recordedRequests()).hasSize(1)
+  }
+
+  @Test
+  fun `should stop retry when getBlockByNumber FINALIED or SAFE is nof found`() {
+    jsonRpcServer.handle("eth_getBlockByNumber") { request ->
+      throw JsonRpcErrorResponseException(rpcErrorCode = -39001, rpcErrorMessage = "Unknown block")
+    }
+    runCatching { client.getBlockByNumberWithoutTransactionsData(BlockParameter.Tag.FINALIZED).get() }
+    assertThat(jsonRpcServer.recordedRequests()).hasSize(1)
+    jsonRpcServer.cleanRecordedRequests()
+
+    runCatching { client.getBlockByNumber(BlockParameter.Tag.FINALIZED).get() }
+    assertThat(jsonRpcServer.recordedRequests()).hasSize(1)
+    jsonRpcServer.cleanRecordedRequests()
+
+    runCatching { client.getBlockByNumberWithoutTransactionsData(BlockParameter.Tag.SAFE).get() }
+    assertThat(jsonRpcServer.recordedRequests()).hasSize(1)
+    jsonRpcServer.cleanRecordedRequests()
+
+    runCatching { client.getBlockByNumber(BlockParameter.Tag.SAFE).get() }
+    assertThat(jsonRpcServer.recordedRequests()).hasSize(1)
+  }
```

Comment:
Can we add test case for `blockParameter` = BlockParameter.Tag.SAFE or BlockParameter.Tag.FINALIZED?

Reply:
I mean test cases for `blockParameter` as `BlockParameter.Tag.LATEST` and to check it doesn't stop retrying on `rpcErrorCode = -39001`